### PR TITLE
Fix address validations

### LINKED ISSUE #2664: Beef up `address` validations
- [x] Prevent port from being added
- [x] Don't require `/` for IP addresses
- [x] Write more tests for edge cases
- [x] No URLs allowed
- [x] Prevent localhost, 127.0.0.1, ::/1, etc, and `0.0.0.0/0`, `::/0`, `100.100.111.1` and the reserved DNS client range from being saved

### PR DESCRIPTION
**Pull request by AndrewDryga**
Based on top of #2670 to avoid introducing conflicts, but it led to GitHub showing more changes than it should, so only look into one last commit.

Closes #2664

### CODE REVIEW

File: elixir/apps/domain/lib/domain/resources/resource/changeset.ex [hunk 1]
```diff
@@ -44,19 +44,63 @@ defmodule Domain.Resources.Resource.Changeset do
...
   end
 
   defp validate_cidr_address(changeset) do
-    changeset = validate_and_normalize_cidr(changeset, :address)
+    changeset
+    |> validate_and_normalize_cidr(:address)
+    |> validate_not_in_cidr(:address, %Postgrex.INET{address: {0, 0, 0, 0}, netmask: 32},
+      message: "can not contain all IPv4 addresses"
+    )
+    |> validate_not_in_cidr(:address, %Postgrex.INET{address: {127, 0, 0, 1}, netmask: 32},
```
Comment (bmanifold) [inline elixir/apps/domain/lib/domain/resources/resource/changeset.ex]: Do we want to make sure that they don't use any loopback interface?  If so, we probably want to use the following:
```
%Postgrex.INET{address: {127, 0, 0, 0}, netmask: 8}
```

File: elixir/apps/domain/lib/domain/resources/resource/changeset.ex [hunk 2]
```diff
@@ -44,19 +44,63 @@ defmodule Domain.Resources.Resource.Changeset do
...
+      },
+      message: "can not contain loopback addresses"
+    )
+    |> validate_address_is_not_in_private_range()
+  end
 
+  defp validate_ip_address(changeset) do
+    changeset
+    |> validate_and_normalize_ip(:address)
+    |> validate_exclusion(:address, ~w[127.0.0.1  ::1],
```
Comment (bmanifold) [inline elixir/apps/domain/lib/domain/resources/resource/changeset.ex]: Similar to my comment above, if we want to prevent any loopback interface from being used, we may need to expand this to exclude any `127.x.x.x.` address

File: elixir/apps/domain/test/domain/resources/resource/changeset_test.exs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      refute create(%{type: :cidr, address: "::8/32"}).valid?
+      refute create(%{type: :cidr, address: "2607:f8b0:4012:0::200e/128:80"}).valid?
+    end
+
+    test "validates and normalizes IP addresses" do
+      for {string, ip} <- [
+            {"192.168.1.1", "192.168.1.1"},
+            {"101.100.100.0", "101.100.100.0"},
+            {"192.168.1.255", "192.168.1.255"},
+            {"192.168.1.255", "192.168.1.255"},
```
Comment (bmanifold) [inline elixir/apps/domain/test/domain/resources/resource/changeset_test.exs]: Just wondering if you intended to have these lines duplicated here?

File: elixir/apps/domain/test/domain/resources/resource/changeset_test.exs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        assert changeset.changes[:address] == ip
+        assert changeset.valid?
+      end
+
+      refute create(%{type: :ip, address: "192.168.1.256"}).valid?
+      refute create(%{type: :ip, address: "100.64.0.0"}).valid?
+      refute create(%{type: :ip, address: "fd00:2021:1111::"}).valid?
+      refute create(%{type: :ip, address: "0.0.0.0"}).valid?
+      refute create(%{type: :ip, address: "0.0.0.0"}).valid?
+      refute create(%{type: :ip, address: "0.0.0.0"}).valid?
```
Comment (bmanifold) [inline elixir/apps/domain/test/domain/resources/resource/changeset_test.exs]: Same as previous question, just wondering about the duplicates here.

**Review by bmanifold — COMMENTED** [review]
Overall it looks good.  Just a couple of small questions

File: elixir/apps/domain/lib/domain/resources/resource/changeset.ex (hunk 1 above)
Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/resources/resource/changeset.ex]: Good catch, I don't think anything other than 127.0.0.1 is supported but let's blacklist them for future.

**Review by bmanifold — APPROVED** [review]
LGTM
